prov/lnx/src/lnx_av.c

```diff
@@ -0,0 +1,702 @@
+
+		/* this is a local peer */
+		lp = calloc(sizeof(*lp), 1);
+		if (!lp)
+			return -FI_ENOMEM;
+
+		dlist_init(&lp->lp_provs);
+
+		rc = is_local_addr(&peer_tbl->lpt_domain->ld_fabric->shm_prov,
+				   la);
+		if (!rc) {
+			lp->lp_local = !disable_shm;
+		} else if (rc == -FI_EOPNOTSUPP) {
+			lp->lp_local = false;
+		} else if (rc) {
+			FI_INFO(&lnx_prov, FI_LOG_CORE, "failed to identify address\n");
+			return rc;
+		}
+
+		rc = lnx_peer_map_addrs(prov_table, lp, la, flags, context);
+		if (rc) {
+			free(lp);
+			return rc;
+		}
+
+		idx = lnx_peer_insert(peer_tbl, lp);
+		if (idx == -1) {
+			rc = lnx_peer_av_remove(lp);
+			lnx_free_peer(lp);
+			FI_INFO(&lnx_prov, FI_LOG_CORE,
+					"Peer table size exceeded. Removed = %d\n", rc);
+			return -FI_ENOENT;
+		}
+
+		fi_addr[i] = (fi_addr_t) idx;
+
+		la = next_peer(la);
+	}
+
+	return i;
+}
+
+int lnx_av_remove(struct fid_av *av, fi_addr_t *fi_addr, size_t count,
+		  uint64_t flags)
+{
+	struct lnx_peer_table *peer_tbl;
+	int frc = 0, rc, i;
+
+	peer_tbl = container_of(av, struct lnx_peer_table, lpt_av.av_fid.fid);
+
+	for (i = 0; i < count; i++) {
+		rc = lnx_peer_remove(peer_tbl, (int)fi_addr[i]);
+		if (rc)
+			frc = rc;
+	}
+
+	return frc;
+}
+
+static const char *
+lnx_av_straddr(struct fid_av *av, const void *addr,
+	       char *buf, size_t *len)
+{
+	/* TODO: implement */
+	return NULL;
+}
+
+static int
+lnx_av_lookup(struct fid_av *av, fi_addr_t fi_addr, void *addr,
+	      size_t *addrlen)
+{
+	/* TODO: implement */
+	return -FI_EOPNOTSUPP;
+}
+
+static struct fi_ops_av lnx_av_ops = {
+	.size = sizeof(struct fi_ops_av),
+	.insert = lnx_av_insert,
+	.remove = lnx_av_remove,
+	.insertsvc = fi_no_av_insertsvc,
+	.insertsym = fi_no_av_insertsym,
+	.lookup = lnx_av_lookup,
+	.straddr = lnx_av_straddr,
+};
+
+static void lnx_get_core_av_attr(struct local_prov_ep *ep,
+				 struct fi_av_attr *attr)
+{
+	memset(attr, 0, sizeof(*attr));
+	attr->type = ep->lpe_fi_info->domain_attr->av_type;
+}
+
+static int lnx_open_avs(struct local_prov *prov, struct fi_av_attr *attr,
+			void *context)
+{
+	int rc = 0;
+	struct local_prov_ep *ep;
+	struct fi_av_attr core_attr;
+
+	dlist_foreach_container(&prov->lpv_prov_eps,
+		struct local_prov_ep, ep, entry) {
+		lnx_get_core_av_attr(ep, &core_attr);
+		if (ep->lpe_local)
+			core_attr.count = ep->lpe_fi_info->domain_attr->ep_cnt;
+		else
+			core_attr.count = attr->count;
+		rc = fi_av_open(ep->lpe_domain, &core_attr,
+					    &ep->lpe_av, context);
+		if (rc)
+			return rc;
+	}
+
+	return 0;
+}
+
+int lnx_av_open(struct fid_domain *domain, struct fi_av_attr *attr,
+		struct fid_av **av, void *context)
+{
+	struct lnx_fabric *fabric;
+	struct lnx_domain *lnx_domain;
+	struct lnx_peer_table *peer_tbl;
+	struct local_prov *entry;
+	size_t table_sz = LNX_DEF_AV_SIZE;
+	int rc = 0;
+
+	if (!attr)
+		return -FI_EINVAL;
+
+	if (attr->name)
+		return -FI_ENOSYS;
+
+	if (attr->type != FI_AV_UNSPEC &&
+	    attr->type != FI_AV_TABLE)
+		return -FI_ENOSYS;
```

Comment:
Technically, an FI_AV_TABLE implementation satisfies the requirements for FI_AV_MAP meaning that lnx can support FI_AV_MAP.